File: Abeyonix_Backend/app/api/v1/inquiry.py
from fastapi import APIRouter, Depends, HTTPException, status, Query as FastQuery
from sqlalchemy import desc
from sqlalchemy.orm import Session, joinedload
from typing import List,Optional

from app.db.base import Base
from app.db.session import get_db  # Your DB session dependency
from app.models.inquiry import Inquiry
from app.schemas.inquiry import InquiryCreate, InquiryResponse


router = APIRouter(
    prefix="/api/v1/inquiry",
    tags=["INQUIRIES"]
)


@router.post("/", response_model=InquiryResponse)
def create_inquiry(payload: InquiryCreate, db: Session = Depends(get_db)):

    inquiry = Inquiry(
        name=payload.name,
        email=payload.email,
        telephone=payload.telephone,
        message=payload.message
    )

    db.add(inquiry)
    db.commit()
    db.refresh(inquiry)

    return inquiry


@router.get("/")
def get_inquiries(
    page: int = FastQuery(1, ge=1),
    page_size: int = FastQuery(10, ge=1, le=100),
    db: Session = Depends(get_db)
):

    # 🔢 Calculate offset
    offset = (page - 1) * page_size

    # 📊 Total count
    total_records = db.query(Inquiry).count()

    # 📦 Fetch paginated data
    inquiries = (
        db.query(Inquiry)
        .order_by(desc(Inquiry.created_at))
        .offset(offset)
        .limit(page_size)
        .all()
    )

    # 🧾 Format response
    records = [
        {
            "id": item.id,
            "name": item.name,
            "email": item.email,
            "telephone": item.telephone,
            "message": item.message,
            "status": item.status,
            "created_at": item.created_at
        }
        for item in inquiries
    ]

    return {
        "total_records": total_records,
        "page": page,
        "page_size": page_size,
        "total_pages": (total_records + page_size - 1) // page_size,
        "records": records
    }

File: Abeyonix_Frontend/src/api/inquiry.ts
import api from "./api";

export interface InquiryPayload {
  name: string;
  email: string;
  telephone?: string;
  message: string;
}

export interface InquiryResponse {
  id: number;
  name: string;
  email: string;
  telephone?: string;
  message: string;
  status: string;
  created_at: string;
}


export interface InquiryListResponse {
  total_records: number;
  page: number;
  page_size: number;
  total_pages: number;
  records: InquiryResponse[];
}



export const createInquiry = async (
  payload: InquiryPayload
): Promise<InquiryResponse> => {
  try {
    const { data } = await api.post<InquiryResponse>(
      "/inquiry/",
      payload
    );

    return data;
  } catch (error: any) {
    throw (
      error?.response?.data?.detail ||
      "Failed to submit inquiry"
    );
  }
};



export const getInquiries = async (params?: {
  page?: number;
  page_size?: number;
}): Promise<InquiryListResponse> => {
  try {
    const { data } = await api.get<InquiryListResponse>(
      "/inquiry/",
      {
        params: {
          page: params?.page || 1,
          page_size: params?.page_size || 10,
        },
      }
    );

    return data;
  } catch (error: any) {
    throw (
      error?.response?.data?.detail ||
      "Failed to fetch inquiries"
    );
  }
};